# Data Cleaning - Batch 4 csv

Majority of the Data Cleaning was done by one of our groupmates named Erwin. 

In the process, I learned a lot and I became more confident of my Data cleaning skills in Python.


In [ ]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

# Library to Display whole Dataset.
pd.set_option("display.max.columns",None)
pd.set_option("display.max.rows",None)

# to suppress warnings
import warnings
warnings.filterwarnings('ignore')


In [2]:
#Pandasai

import os
from pandasai import PandasAI
from pandasai.llm.openai import OpenAI

openai_api_key = "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"
llm = OpenAI(api_token=openai_api_key)
pandas_ai = PandasAI(llm)

# Data Dictionary


|STT | Attribute Name | Meaning |
|:---| :--- |:--- |
| 1 | CaseCode | Unique Case ID given to the patient |
| 2 | Age | Patient's Age |
| 3 | AgeGroup | Age Group the patient belongs to |
| 4 | Sex | Biological Sex |
| 5 | DateSpecimen | Date when specimen was taken from the patient |
| 6 | DateResultRelease | Date when specimen was released |
| 7 | DateRepConf | Confirmation date that the patient has contracted COVID-19 |
| 8 | DateDied | Date of death |
| 9 | DateRecover | Date of recovery |
| 10 | RemovalType | Patient status: Recovered or Died |
| 11 | Admitted | Refers to if the patient was admitted in the hospital or not |
| 12 | RegionRes | Region where the patient is residing |
| 13 | ProvRes | Province where the patient is residing |
| 14 | CityMunRes | City/Municipality where the patient is residing |
| 15 | CityMuniPSGC | Refers to the Philippine Standard Geographic Code of the City/Municipality |
| 16 | BarangayRes | Barangay where the patient is residing |
| 17 | BarangayPSGC | Philippine Standard Geographic Code of the Barangay |
| 18 | HealthStatus | Patient status: Recovered or Died *redundant, this column is same with RemovalType* |
| 19 | Quarantined | If the patient was quarantined or not |
| 20 | DateOnset | First clinical symptom or sign for a particular condition *definition taken from google* |
| 21 | Pregnanttab | If the patient was pregnant or not |
| 22 | ValidationStatus | Additional insights on each patient particularly on missing or confusing information |


In [2]:
#read csv file in Capstone folder
batch4 = pd.read_csv("DOH COVID Data Drop_ 20230617 - 04 Case Information_batch_4.csv")


batch4.head()

CaseCode   Age  AgeGroup     Sex DateSpecimen DateResultRelease  \
0  C38887343  59.0  55 to 59    MALE          NaN               NaN   
1  C38070221  26.0  25 to 29  FEMALE   2022-10-28        2022-10-30   
2  C23752209  58.0  55 to 59    MALE   2022-10-28        2022-10-29   
3  C73020737  32.0  30 to 34    MALE          NaN               NaN   
4  C72721234  31.0  30 to 34  FEMALE   2022-10-28        2022-10-30   

  DateRepConf DateDied DateRecover RemovalType Admitted  \
0  2022-11-01      NaN         NaN   RECOVERED       NO   
1  2022-11-01      NaN         NaN   RECOVERED       NO   
2  2022-11-01      NaN         NaN   RECOVERED       NO   
3  2022-11-01      NaN         NaN   RECOVERED       NO   
4  2022-11-01      NaN         NaN   RECOVERED       NO   

                     RegionRes          ProvRes                CityMunRes  \
0       Region V: Bicol Region    CAMARINES SUR                       NaN   
1  Region VII: Central Visayas             CEBU       CEBU CITY (CAPITAL)   
2  Region VII: Central Visayas  NEGROS ORIENTAL  DUMAGUETE CITY (CAPITAL)   
3        Region IV-B: MIMAROPA       MARINDUQUE                  TORRIJOS   
4    Region II: Cagayan Valley    NUEVA VIZCAYA                    KAYAPA   

  CityMuniPSGC   BarangayRes BarangayPSGC HealthStatus Quarantined  \
0          NaN           NaN          NaN    RECOVERED          NO   
1  PH072217000  LAHUG (POB.)  PH072217041    RECOVERED          NO   
2  PH074610000          DARO  PH074610013    RECOVERED          NO   
3  PH174006000          SUHA  PH174006023    RECOVERED          NO   
4  PH025010000        BESONG  PH025010009    RECOVERED          NO   

    DateOnset Pregnanttab                                   ValidationStatus  
0  2022-10-29         NaN  Removal Type is "Recovered", but no Recovered ...  
1  2022-10-28          NO  Health Status is "Recovered", but no Date Reco...  
2         NaN         NaN  Removal Type is "Recovered", but no Recovered ...  
3         NaN         NaN  Removal Type is "Recovered", but no Recovered ...  
4  2022-10-26          NO  Removal Type is "Recovered", but no Recovered ...

### Filter: include only "City of Manila"

In [3]:
#filtering the dataset to get only the total cases for the City of Manila.
batch4_manila = batch4[batch4["CityMunRes"] == "CITY OF MANILA"]

In [4]:
batch4_manila.shape

(7287, 22)

In [5]:
batch4_manila.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7287 entries, 17 to 158543
Data columns (total 22 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CaseCode           7287 non-null   object 
 1   Age                7221 non-null   float64
 2   AgeGroup           7221 non-null   object 
 3   Sex                7287 non-null   object 
 4   DateSpecimen       6359 non-null   object 
 5   DateResultRelease  6377 non-null   object 
 6   DateRepConf        7287 non-null   object 
 7   DateDied           9 non-null      object 
 8   DateRecover        928 non-null    object 
 9   RemovalType        7103 non-null   object 
 10  Admitted           3814 non-null   object 
 11  RegionRes          7287 non-null   object 
 12  ProvRes            7287 non-null   object 
 13  CityMunRes         7287 non-null   object 
 14  CityMuniPSGC       7287 non-null   object 
 15  BarangayRes        7234 non-null   object 
 16  BarangayPSGC       72

In [6]:
batch4_manila.describe()

Age
count  7221.000000
mean     37.817061
std      16.556783
min       0.000000
25%      27.000000
50%      34.000000
75%      47.000000
max     104.000000

# Checking for null values

In [7]:
batch4_manila.isnull().sum()

CaseCode                0
Age                    66
AgeGroup               66
Sex                     0
DateSpecimen          928
DateResultRelease     910
DateRepConf             0
DateDied             7278
DateRecover          6359
RemovalType           184
Admitted             3473
RegionRes               0
ProvRes                 0
CityMunRes              0
CityMuniPSGC            0
BarangayRes            53
BarangayPSGC           53
HealthStatus            0
Quarantined          2852
DateOnset            5819
Pregnanttab          5098
ValidationStatus      994
dtype: int64

In [8]:
batch4_manila.isnull().sum().to_frame()

0
CaseCode              0
Age                  66
AgeGroup             66
Sex                   0
DateSpecimen        928
DateResultRelease   910
DateRepConf           0
DateDied           7278
DateRecover        6359
RemovalType         184
Admitted           3473
RegionRes             0
ProvRes               0
CityMunRes            0
CityMuniPSGC          0
BarangayRes          53
BarangayPSGC         53
HealthStatus          0
Quarantined        2852
DateOnset          5819
Pregnanttab        5098
ValidationStatus    994

In [9]:
## This is a better way to explain it to SMEs and stakeholders


missing_batch4 = batch4_manila.isnull().sum().to_frame().rename(columns={0:"Total No. of Missing Values"})

# Divided the total No. of Missing values by the Total length of the dataframe, then round it to 2 decimal places
missing_batch4["% of Missing Values"] = round((missing_batch4["Total No. of Missing Values"]/len(batch4_manila))*100,2)






missing_batch4

Total No. of Missing Values  % of Missing Values
CaseCode                                     0                 0.00
Age                                         66                 0.91
AgeGroup                                    66                 0.91
Sex                                          0                 0.00
DateSpecimen                               928                12.74
DateResultRelease                          910                12.49
DateRepConf                                  0                 0.00
DateDied                                  7278                99.88
DateRecover                               6359                87.26
RemovalType                                184                 2.53
Admitted                                  3473                47.66
RegionRes                                    0                 0.00
ProvRes                                      0                 0.00
CityMunRes                                   0                 0.00
CityMuniPSGC                                 0                 0.00
BarangayRes                                 53                 0.73
BarangayPSGC                                53                 0.73
HealthStatus                                 0                 0.00
Quarantined                               2852                39.14
DateOnset                                 5819                79.85
Pregnanttab                               5098                69.96
ValidationStatus                           994                13.64

In [10]:

#missing1 = batch1.isnull().sum().to_frame().rename(columns={0:"Total No. of Missing Values"})


# Divided the total No. of Missing values by the Total length of the dataframe, then round it to 2 decimal places
#missing1["% of Missing Values"] = round((missing1["Total No. of Missing Values"]/len(batch1))*100,2)


#missing1

### Checking the Pregnanttab

In [11]:
batch4_manila['Pregnanttab'].unique()


array([nan, 'NO'], dtype=object)

In [12]:
batch4_manila['Pregnanttab'].value_counts()

NO    2189
Name: Pregnanttab, dtype: int64

In [13]:
batch4_manila['Sex'].value_counts()

MALE      3778
FEMALE    3509
Name: Sex, dtype: int64

Inferences:

1. High percentage of missing values  for the following columns: DateDied, DateRecover, DateOnset, DateSpecimen, DateResultRelease
2. Upon checking, the missing values for the Pregnanttab, they are actually not missing values, rather these are the total number of males in the dataset which were given NaN values.

<br><br><br><br><br>

# Descriptive Statistics

- Numerical
- Categorical

### Numerical

Note:

- Count: It shows the number of values that are not missing in each of the column.<br>
- Mean : It shows the average of the values in each of the column.<br>
- Std :  It shows standard deviation of the values in each of the column.<br>
- Min :  It shows the smallest value in each of the column.<br>
- Max :  It shows the maximum value in each of the column.<br>
- 25%, 50%, 75%: The valuescorrespond to the data divisions of 25%, 50% and 75% percentiles.<br></div>

In [14]:
batch4_manila.describe().T

count       mean        std  min   25%   50%   75%    max
Age  7221.0  37.817061  16.556783  0.0  27.0  34.0  47.0  104.0

Inferences:

1. There are 3 missing values in the Age column
2. The mean age is approximately 38, which means we can use this value to fill in the other missing values to help our analysis.

### Categorical

Note:

- Count: It shows the number of values that are not missing in each of the column.<br>
- Unique : It shows the number of unique values present in each of the ccolumn.<br>
- Top : It shows which categorical value is present most in each of the column.<br>
- Freq : It shows the frequency of the categorical value apperaing most in each of the columns.<br></div>

In [15]:
batch4_manila.describe(include="O").T

count unique  \
CaseCode           7287   7287   
AgeGroup           7221     17   
Sex                7287      2   
DateSpecimen       6359    284   
DateResultRelease  6377    280   
DateRepConf        7287    229   
DateDied              9      9   
DateRecover         928    180   
RemovalType        7103      2   
Admitted           3814      2   
RegionRes          7287      1   
ProvRes            7287      1   
CityMunRes         7287      1   
CityMuniPSGC       7287     15   
BarangayRes        7234    861   
BarangayPSGC       7234    861   
HealthStatus       7287      7   
Quarantined        4435      2   
DateOnset          1468    254   
Pregnanttab        2189      1   
ValidationStatus   6293    182   

                                                                 top  freq  
CaseCode                                                   C16163369     1  
AgeGroup                                                    25 to 29  1376  
Sex                                                             MALE  3778  
DateSpecimen                                              2023-05-11   104  
DateResultRelease                                         2023-05-11   107  
DateRepConf                                               2023-05-06   106  
DateDied                                                  2022-11-26     1  
DateRecover                                               2022-12-08    24  
RemovalType                                                RECOVERED  7094  
Admitted                                                          NO  3811  
RegionRes                                                        NCR  7287  
ProvRes                                                          NCR  7287  
CityMunRes                                            CITY OF MANILA  7287  
CityMuniPSGC                                             PH133906000  1633  
BarangayRes                                             BARANGAY 830   201  
BarangayPSGC                                             PH133911045   201  
HealthStatus                                               RECOVERED  7094  
Quarantined                                                       NO  4313  
DateOnset                                                 2023-05-04    24  
Pregnanttab                                                       NO  2189  
ValidationStatus   Health Status is "Recovered", but no Date Reco...  2930

# Inference:

1. CaseCode will not be useful for the analysis since it is only a unique code given to each patient.
2. The most common/frequent age group in the data is between 25 to 29.
3. RemovalType and HealthStatus contains similar values, but upon cross-checking with other batches of DOH's data, HealthStatus also contains other values.
4. Pregnanttab will not be useful for the analysis. DOH recorded the males as NaN and for this dataset all the females were not pregnant.

<br>
<br>

In [16]:
batch4_manila['Sex'].value_counts()


MALE      3778
FEMALE    3509
Name: Sex, dtype: int64

## Creating a Copy of the dataset 
## and Dropping Unnecessary Columns

In [17]:
#make a copy of the dataset
batch4_manila_copy = batch4_manila.copy()

In [18]:
# I also dropped CaseCode and RemovalType
cols = ['CaseCode','CityMuniPSGC', 'BarangayPSGC']
batch4_manila_copy.drop(columns=cols, inplace=True)

### AgeGroup Column:

Insights:
1. Upon checking the percentage of missing values for each column in the dataset, we decided to drop the rows with missing values for the 'AgeGroup' Column because it made up less than 2% of the dataset.

In [19]:
#drop missing values in the AgeGroup column.
batch4_manila_copy.dropna(subset=['AgeGroup'], inplace=True)

In [20]:
batch4_manila_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7221 entries, 17 to 158543
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Age                7221 non-null   float64
 1   AgeGroup           7221 non-null   object 
 2   Sex                7221 non-null   object 
 3   DateSpecimen       6295 non-null   object 
 4   DateResultRelease  6314 non-null   object 
 5   DateRepConf        7221 non-null   object 
 6   DateDied           9 non-null      object 
 7   DateRecover        911 non-null    object 
 8   RemovalType        7037 non-null   object 
 9   Admitted           3762 non-null   object 
 10  RegionRes          7221 non-null   object 
 11  ProvRes            7221 non-null   object 
 12  CityMunRes         7221 non-null   object 
 13  BarangayRes        7169 non-null   object 
 14  HealthStatus       7221 non-null   object 
 15  Quarantined        4369 non-null   object 
 16  DateOnset          14

In [21]:
batch4_manila_copy.head()

Age  AgeGroup     Sex DateSpecimen DateResultRelease DateRepConf  \
17   24.0  20 to 24    MALE          NaN               NaN  2022-11-01   
97   26.0  25 to 29    MALE          NaN               NaN  2022-11-01   
104  54.0  50 to 54    MALE          NaN               NaN  2022-11-01   
121  36.0  35 to 39  FEMALE   2022-10-28        2022-10-29  2022-11-01   
148  22.0  20 to 24    MALE   2022-10-29        2022-10-29  2022-11-01   

    DateDied DateRecover RemovalType Admitted RegionRes ProvRes  \
17       NaN         NaN   RECOVERED       NO       NCR     NCR   
97       NaN         NaN   RECOVERED       NO       NCR     NCR   
104      NaN         NaN   RECOVERED       NO       NCR     NCR   
121      NaN         NaN   RECOVERED       NO       NCR     NCR   
148      NaN         NaN   RECOVERED       NO       NCR     NCR   

         CityMunRes   BarangayRes HealthStatus Quarantined DateOnset  \
17   CITY OF MANILA  BARANGAY 675    RECOVERED          NO       NaN   
97   CITY OF MANILA  BARANGAY 459    RECOVERED          NO       NaN   
104  CITY OF MANILA  BARANGAY 318    RECOVERED          NO       NaN   
121  CITY OF MANILA  BARANGAY 179    RECOVERED          NO       NaN   
148  CITY OF MANILA  BARANGAY 654    RECOVERED          NO       NaN   

    Pregnanttab                                   ValidationStatus  
17          NaN  Health Status is "Recovered", but no Date Reco...  
97          NaN  Health Status is "Recovered", but no Date Reco...  
104         NaN  Health Status is "Recovered", but no Date Reco...  
121          NO  Removal Type is "Recovered", but no Recovered ...  
148         NaN  Health Status is "Recovered", but no Date Reco...

In [22]:
missing_batch4_manila_copy = batch4_manila_copy.isnull().sum().to_frame().rename(columns={0:"Total No. of Missing Values"})


# Divided the total No. of Missing values by the Total length of the dataframe, then round it to 2 decimal places
missing_batch4_manila_copy["% of Missing Values"] = round((missing_batch4_manila_copy["Total No. of Missing Values"]/len(batch4_manila_copy))*100,2)





missing_batch4_manila_copy

Total No. of Missing Values  % of Missing Values
Age                                          0                 0.00
AgeGroup                                     0                 0.00
Sex                                          0                 0.00
DateSpecimen                               926                12.82
DateResultRelease                          907                12.56
DateRepConf                                  0                 0.00
DateDied                                  7212                99.88
DateRecover                               6310                87.38
RemovalType                                184                 2.55
Admitted                                  3459                47.90
RegionRes                                    0                 0.00
ProvRes                                      0                 0.00
CityMunRes                                   0                 0.00
BarangayRes                                 52                 0.72
HealthStatus                                 0                 0.00
Quarantined                               2852                39.50
DateOnset                                 5783                80.09
Pregnanttab                               5060                70.07
ValidationStatus                           977                13.53

### Check Unique Values of Categorical Attributes.

In [23]:
unq_columns = batch4_manila_copy.select_dtypes(include="O").columns      #selects the columns of data types that are objects



unq_columns

Index(['AgeGroup', 'Sex', 'DateSpecimen', 'DateResultRelease', 'DateRepConf',
       'DateDied', 'DateRecover', 'RemovalType', 'Admitted', 'RegionRes',
       'ProvRes', 'CityMunRes', 'BarangayRes', 'HealthStatus', 'Quarantined',
       'DateOnset', 'Pregnanttab', 'ValidationStatus'],
      dtype='object')

In [24]:
#not sure if this is needed, I was just checking notes from the lectures

def preprocess(ValidationStatus):
    ValidationStatus = ValidationStatus.str.replace("(<br/>)", "")
    ValidationStatus = ValidationStatus.str.replace('(<a).*(>).*(</a>)', '')
    ValidationStatus = ValidationStatus.str.replace('(&amp)', '')
    ValidationStatus = ValidationStatus.str.replace('(&gt)', '')
    ValidationStatus = ValidationStatus.str.replace('(&lt)', '')
    ValidationStatus = ValidationStatus.str.replace('(\xa0)', ' ') 
    ValidationStatus = ValidationStatus.str.replace('(\n)', ' ') 
    return ValidationStatus

batch4_manila_copy['ValidationStatus'] = preprocess(batch4_manila_copy['ValidationStatus'])

In [25]:
unq_columns = batch4_manila_copy.select_dtypes(include="O").columns

for column in unq_columns:
    print('Unique values of ', column, set(batch4_manila_copy[column]))         # set is used to place the unique values in a set
    print("-"*127)

Unique values of  AgeGroup {'40 to 44', '45 to 49', '60 to 64', '55 to 59', '50 to 54', '75 to 79', '20 to 24', '0 to 4', '80+', '30 to 34', '70 to 74', '65 to 69', '5 to 9', '10 to 14', '15 to 19', '25 to 29', '35 to 39'}
-------------------------------------------------------------------------------------------------------------------------------
Unique values of  Sex {'FEMALE', 'MALE'}
-------------------------------------------------------------------------------------------------------------------------------
Unique values of  DateSpecimen {'2023-02-27', '2023-01-24', '2023-06-09', '2022-12-05', '2022-12-09', '2022-07-13', '2022-12-10', '2023-06-01', '2022-11-07', '2023-06-04', '2023-05-29', '2022-10-01', '2022-09-13', '2022-03-23', '2023-01-06', '2022-12-04', '2023-01-25', '2023-04-09', '2023-02-16', '2023-03-28', '2021-04-13', '2022-11-03', '2022-01-12', '2022-10-16', '2023-05-05', '2023-01-18', '2023-03-31', '2023-05-11', '2023-02-15', '2022-11-22', '2022-12-08', '2022-06-11', 

In [26]:
batch4_manila_copy.head()

Age  AgeGroup     Sex DateSpecimen DateResultRelease DateRepConf  \
17   24.0  20 to 24    MALE          NaN               NaN  2022-11-01   
97   26.0  25 to 29    MALE          NaN               NaN  2022-11-01   
104  54.0  50 to 54    MALE          NaN               NaN  2022-11-01   
121  36.0  35 to 39  FEMALE   2022-10-28        2022-10-29  2022-11-01   
148  22.0  20 to 24    MALE   2022-10-29        2022-10-29  2022-11-01   

    DateDied DateRecover RemovalType Admitted RegionRes ProvRes  \
17       NaN         NaN   RECOVERED       NO       NCR     NCR   
97       NaN         NaN   RECOVERED       NO       NCR     NCR   
104      NaN         NaN   RECOVERED       NO       NCR     NCR   
121      NaN         NaN   RECOVERED       NO       NCR     NCR   
148      NaN         NaN   RECOVERED       NO       NCR     NCR   

         CityMunRes   BarangayRes HealthStatus Quarantined DateOnset  \
17   CITY OF MANILA  BARANGAY 675    RECOVERED          NO       NaN   
97   CITY OF MANILA  BARANGAY 459    RECOVERED          NO       NaN   
104  CITY OF MANILA  BARANGAY 318    RECOVERED          NO       NaN   
121  CITY OF MANILA  BARANGAY 179    RECOVERED          NO       NaN   
148  CITY OF MANILA  BARANGAY 654    RECOVERED          NO       NaN   

    Pregnanttab                                   ValidationStatus  
17          NaN  Health Status is "Recovered", but no Date Reco...  
97          NaN  Health Status is "Recovered", but no Date Reco...  
104         NaN  Health Status is "Recovered", but no Date Reco...  
121          NO  Removal Type is "Recovered", but no Recovered ...  
148         NaN  Health Status is "Recovered", but no Date Reco...

In [27]:
batch4_manila_copy.shape

(7221, 19)

## In the Pregnanttab column, fill null values with no.


In [28]:
# In the Pregnanttab column, fill null values with no.

batch4_manila_copy['Pregnanttab'].fillna('No', inplace=True)

In [29]:
batch4_manila_copy.head()

Age  AgeGroup     Sex DateSpecimen DateResultRelease DateRepConf  \
17   24.0  20 to 24    MALE          NaN               NaN  2022-11-01   
97   26.0  25 to 29    MALE          NaN               NaN  2022-11-01   
104  54.0  50 to 54    MALE          NaN               NaN  2022-11-01   
121  36.0  35 to 39  FEMALE   2022-10-28        2022-10-29  2022-11-01   
148  22.0  20 to 24    MALE   2022-10-29        2022-10-29  2022-11-01   

    DateDied DateRecover RemovalType Admitted RegionRes ProvRes  \
17       NaN         NaN   RECOVERED       NO       NCR     NCR   
97       NaN         NaN   RECOVERED       NO       NCR     NCR   
104      NaN         NaN   RECOVERED       NO       NCR     NCR   
121      NaN         NaN   RECOVERED       NO       NCR     NCR   
148      NaN         NaN   RECOVERED       NO       NCR     NCR   

         CityMunRes   BarangayRes HealthStatus Quarantined DateOnset  \
17   CITY OF MANILA  BARANGAY 675    RECOVERED          NO       NaN   
97   CITY OF MANILA  BARANGAY 459    RECOVERED          NO       NaN   
104  CITY OF MANILA  BARANGAY 318    RECOVERED          NO       NaN   
121  CITY OF MANILA  BARANGAY 179    RECOVERED          NO       NaN   
148  CITY OF MANILA  BARANGAY 654    RECOVERED          NO       NaN   

    Pregnanttab                                   ValidationStatus  
17           No  Health Status is "Recovered", but no Date Reco...  
97           No  Health Status is "Recovered", but no Date Reco...  
104          No  Health Status is "Recovered", but no Date Reco...  
121          NO  Removal Type is "Recovered", but no Recovered ...  
148          No  Health Status is "Recovered", but no Date Reco...

### Dealing with DateSpecimen, DateResultRelease, DateDied, DateRecover, DateOnset

In [30]:
missing_batch4_manila_copy

Total No. of Missing Values  % of Missing Values
Age                                          0                 0.00
AgeGroup                                     0                 0.00
Sex                                          0                 0.00
DateSpecimen                               926                12.82
DateResultRelease                          907                12.56
DateRepConf                                  0                 0.00
DateDied                                  7212                99.88
DateRecover                               6310                87.38
RemovalType                                184                 2.55
Admitted                                  3459                47.90
RegionRes                                    0                 0.00
ProvRes                                      0                 0.00
CityMunRes                                   0                 0.00
BarangayRes                                 52                 0.72
HealthStatus                                 0                 0.00
Quarantined                               2852                39.50
DateOnset                                 5783                80.09
Pregnanttab                               5060                70.07
ValidationStatus                           977                13.53

In [31]:
batch4_manila_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7221 entries, 17 to 158543
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Age                7221 non-null   float64
 1   AgeGroup           7221 non-null   object 
 2   Sex                7221 non-null   object 
 3   DateSpecimen       6295 non-null   object 
 4   DateResultRelease  6314 non-null   object 
 5   DateRepConf        7221 non-null   object 
 6   DateDied           9 non-null      object 
 7   DateRecover        911 non-null    object 
 8   RemovalType        7037 non-null   object 
 9   Admitted           3762 non-null   object 
 10  RegionRes          7221 non-null   object 
 11  ProvRes            7221 non-null   object 
 12  CityMunRes         7221 non-null   object 
 13  BarangayRes        7169 non-null   object 
 14  HealthStatus       7221 non-null   object 
 15  Quarantined        4369 non-null   object 
 16  DateOnset          14

## Fill the missing values in DateSpecimen, DateResultRelease, DateDied, DateRecover and DateOnset with 'Not Recorded'


In [32]:
#Fill the missing values in DateSpecimen, DateResultRelease, DateDied, DateRecover and DateOnset with 'Not Recorded'
batch4_manila_copy['DateSpecimen'].fillna('Not Recorded', inplace=True)

batch4_manila_copy['DateResultRelease'].fillna('Not Recorded', inplace=True)

batch4_manila_copy['DateDied'].fillna('Not Recorded', inplace=True)

batch4_manila_copy['DateRecover'].fillna('Not Recorded', inplace=True)

batch4_manila_copy['DateOnset'].fillna('Not Recorded', inplace=True)


In [33]:
batch4_manila_copy.head()

Age  AgeGroup     Sex  DateSpecimen DateResultRelease DateRepConf  \
17   24.0  20 to 24    MALE  Not Recorded      Not Recorded  2022-11-01   
97   26.0  25 to 29    MALE  Not Recorded      Not Recorded  2022-11-01   
104  54.0  50 to 54    MALE  Not Recorded      Not Recorded  2022-11-01   
121  36.0  35 to 39  FEMALE    2022-10-28        2022-10-29  2022-11-01   
148  22.0  20 to 24    MALE    2022-10-29        2022-10-29  2022-11-01   

         DateDied   DateRecover RemovalType Admitted RegionRes ProvRes  \
17   Not Recorded  Not Recorded   RECOVERED       NO       NCR     NCR   
97   Not Recorded  Not Recorded   RECOVERED       NO       NCR     NCR   
104  Not Recorded  Not Recorded   RECOVERED       NO       NCR     NCR   
121  Not Recorded  Not Recorded   RECOVERED       NO       NCR     NCR   
148  Not Recorded  Not Recorded   RECOVERED       NO       NCR     NCR   

         CityMunRes   BarangayRes HealthStatus Quarantined     DateOnset  \
17   CITY OF MANILA  BARANGAY 675    RECOVERED          NO  Not Recorded   
97   CITY OF MANILA  BARANGAY 459    RECOVERED          NO  Not Recorded   
104  CITY OF MANILA  BARANGAY 318    RECOVERED          NO  Not Recorded   
121  CITY OF MANILA  BARANGAY 179    RECOVERED          NO  Not Recorded   
148  CITY OF MANILA  BARANGAY 654    RECOVERED          NO  Not Recorded   

    Pregnanttab                                   ValidationStatus  
17           No  Health Status is "Recovered", but no Date Reco...  
97           No  Health Status is "Recovered", but no Date Reco...  
104          No  Health Status is "Recovered", but no Date Reco...  
121          NO  Removal Type is "Recovered", but no Recovered ...  
148          No  Health Status is "Recovered", but no Date Reco...

# Cleaning BarangayRes and ValidationStatus Columns

In [34]:
#Count the total number of missing values in the BarangayRes column.

batch4_manila_copy['BarangayRes'].isnull().value_counts().to_frame().rename(columns={'BarangayRes':'No. of Missing Values'})

No. of Missing Values
False                   7169
True                      52

In [35]:
batch4_manila_copy['BarangayRes'].value_counts().to_frame().rename(columns={'BarangayRes':'No. of Cases'}).sort_values(by='No. of Cases', ascending=False).head(10)

No. of Cases
BARANGAY 830           201
BARANGAY 719           161
BARANGAY 667           143
BARANGAY 666           107
BARANGAY 669           107
BARANGAY 311           106
BARANGAY 670           101
BARANGAY 699            88
BARANGAY 668            72
BARANGAY 697            70

## Fill in the missing values in BarangayRes with 'Not Recorded'.


In [36]:
#fill in the missing values in BarangayRes with 'Not Recorded'.
batch4_manila_copy['BarangayRes'].fillna('Not Recorded', inplace=True)

In [37]:
batch4_manila_copy.head()

Age  AgeGroup     Sex  DateSpecimen DateResultRelease DateRepConf  \
17   24.0  20 to 24    MALE  Not Recorded      Not Recorded  2022-11-01   
97   26.0  25 to 29    MALE  Not Recorded      Not Recorded  2022-11-01   
104  54.0  50 to 54    MALE  Not Recorded      Not Recorded  2022-11-01   
121  36.0  35 to 39  FEMALE    2022-10-28        2022-10-29  2022-11-01   
148  22.0  20 to 24    MALE    2022-10-29        2022-10-29  2022-11-01   

         DateDied   DateRecover RemovalType Admitted RegionRes ProvRes  \
17   Not Recorded  Not Recorded   RECOVERED       NO       NCR     NCR   
97   Not Recorded  Not Recorded   RECOVERED       NO       NCR     NCR   
104  Not Recorded  Not Recorded   RECOVERED       NO       NCR     NCR   
121  Not Recorded  Not Recorded   RECOVERED       NO       NCR     NCR   
148  Not Recorded  Not Recorded   RECOVERED       NO       NCR     NCR   

         CityMunRes   BarangayRes HealthStatus Quarantined     DateOnset  \
17   CITY OF MANILA  BARANGAY 675    RECOVERED          NO  Not Recorded   
97   CITY OF MANILA  BARANGAY 459    RECOVERED          NO  Not Recorded   
104  CITY OF MANILA  BARANGAY 318    RECOVERED          NO  Not Recorded   
121  CITY OF MANILA  BARANGAY 179    RECOVERED          NO  Not Recorded   
148  CITY OF MANILA  BARANGAY 654    RECOVERED          NO  Not Recorded   

    Pregnanttab                                   ValidationStatus  
17           No  Health Status is "Recovered", but no Date Reco...  
97           No  Health Status is "Recovered", but no Date Reco...  
104          No  Health Status is "Recovered", but no Date Reco...  
121          NO  Removal Type is "Recovered", but no Recovered ...  
148          No  Health Status is "Recovered", but no Date Reco...

## **Count the total number of missing values in the ValidationStatus column.**


In [38]:
#Count the total number of missing values in the ValidationStatus column.

batch4_manila_copy['ValidationStatus'].isnull().value_counts().to_frame().rename(columns={'ValidationStatus':'No. of Missing Values'})

No. of Missing Values
False                   6244
True                     977

In [39]:
batch4_manila_copy['ValidationStatus'].value_counts().to_frame().rename(columns={'ValidationStatus':'No. of Cases'}).sort_values(by='No. of Cases', ascending=False).head(10)

No. of Cases
Health Status is "Recovered", but no Date Recov...          2906
Removal Type is "Recovered", but no Recovered D...          2786
Age or Birthdate is Invalid Removal Type is "Re...            32
Age or Birthdate is Invalid Health Status is "R...            28
Case has Lab Result, but Result Date is blank                 25
Age or Birthdate is Invalid                                   25
Removal Type is "Recovered", but no Recovered D...            23
Health Status is "Recovered", but no Date Recov...            23
Health Status is "Recovered", but no Date Recov...            22
Health Status is "Recovered", but no Date Recov...            20

## Fill in the missing values in the ValidationStatus with 'No Comments'.


In [40]:
#fill in the missing values in the ValidationStatus with 'No Comments'.
batch4_manila_copy['ValidationStatus'].fillna('No Comments', inplace=True)

In [41]:
batch4_manila_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7221 entries, 17 to 158543
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Age                7221 non-null   float64
 1   AgeGroup           7221 non-null   object 
 2   Sex                7221 non-null   object 
 3   DateSpecimen       7221 non-null   object 
 4   DateResultRelease  7221 non-null   object 
 5   DateRepConf        7221 non-null   object 
 6   DateDied           7221 non-null   object 
 7   DateRecover        7221 non-null   object 
 8   RemovalType        7037 non-null   object 
 9   Admitted           3762 non-null   object 
 10  RegionRes          7221 non-null   object 
 11  ProvRes            7221 non-null   object 
 12  CityMunRes         7221 non-null   object 
 13  BarangayRes        7221 non-null   object 
 14  HealthStatus       7221 non-null   object 
 15  Quarantined        4369 non-null   object 
 16  DateOnset          72

In [42]:

missing_batch4_manila_copy

Total No. of Missing Values  % of Missing Values
Age                                          0                 0.00
AgeGroup                                     0                 0.00
Sex                                          0                 0.00
DateSpecimen                               926                12.82
DateResultRelease                          907                12.56
DateRepConf                                  0                 0.00
DateDied                                  7212                99.88
DateRecover                               6310                87.38
RemovalType                                184                 2.55
Admitted                                  3459                47.90
RegionRes                                    0                 0.00
ProvRes                                      0                 0.00
CityMunRes                                   0                 0.00
BarangayRes                                 52                 0.72
HealthStatus                                 0                 0.00
Quarantined                               2852                39.50
DateOnset                                 5783                80.09
Pregnanttab                               5060                70.07
ValidationStatus                           977                13.53

# Columns with null values left:

- RemovalType
- Admitted
- Quarantined

<br>
<br>

## Make a 2nd copy of the dataset

In [43]:
#make a 2nd copy of the dataset
batch4_manila_copy2 = batch4_manila_copy.copy()

## Fill in the missing values in RemovalType with 'Not Recorded'.


In [44]:
#fill in the missing values in RemovalType with 'Not Recorded'.
batch4_manila_copy2['RemovalType'].fillna('Not Recorded', inplace=True)

In [45]:
missing_batch4_manila_copy2 = batch4_manila_copy2.isnull().sum().to_frame().rename(columns={0:"Total No. of Missing Values"})


# Divided the total No. of Missing values by the Total length of the dataframe, then round it to 2 decimal places
missing_batch4_manila_copy2["% of Missing Values"] = round((missing_batch4_manila_copy2["Total No. of Missing Values"]/len(batch4_manila_copy2))*100,2)





missing_batch4_manila_copy2

Total No. of Missing Values  % of Missing Values
Age                                          0                  0.0
AgeGroup                                     0                  0.0
Sex                                          0                  0.0
DateSpecimen                                 0                  0.0
DateResultRelease                            0                  0.0
DateRepConf                                  0                  0.0
DateDied                                     0                  0.0
DateRecover                                  0                  0.0
RemovalType                                  0                  0.0
Admitted                                  3459                 47.9
RegionRes                                    0                  0.0
ProvRes                                      0                  0.0
CityMunRes                                   0                  0.0
BarangayRes                                  0                  0.0
HealthStatus                                 0                  0.0
Quarantined                               2852                 39.5
DateOnset                                    0                  0.0
Pregnanttab                                  0                  0.0
ValidationStatus                             0                  0.0

## Fill in the missing values in Admitted with 'Not Recorded'.


In [46]:
#fill in the missing values in Admitted with 'Not Recorded'.
batch4_manila_copy2['Admitted'].fillna('Not Recorded', inplace=True)

## Fill in the missing values in Quarantined with 'Not Recorded'.


In [47]:
#fill in the missing values in Quarantined with 'Not Recorded'.
batch4_manila_copy2['Quarantined'].fillna('Not Recorded', inplace=True)

# CHECKING FOR NULL VALUES:

In [48]:
missing_batch4_manila_copy2

# ?? maybe just an error, if i use .info() it shows that all rows are non-null already.

Total No. of Missing Values  % of Missing Values
Age                                          0                  0.0
AgeGroup                                     0                  0.0
Sex                                          0                  0.0
DateSpecimen                                 0                  0.0
DateResultRelease                            0                  0.0
DateRepConf                                  0                  0.0
DateDied                                     0                  0.0
DateRecover                                  0                  0.0
RemovalType                                  0                  0.0
Admitted                                  3459                 47.9
RegionRes                                    0                  0.0
ProvRes                                      0                  0.0
CityMunRes                                   0                  0.0
BarangayRes                                  0                  0.0
HealthStatus                                 0                  0.0
Quarantined                               2852                 39.5
DateOnset                                    0                  0.0
Pregnanttab                                  0                  0.0
ValidationStatus                             0                  0.0

In [49]:
# FINAL

batch4_manila_copy2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7221 entries, 17 to 158543
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Age                7221 non-null   float64
 1   AgeGroup           7221 non-null   object 
 2   Sex                7221 non-null   object 
 3   DateSpecimen       7221 non-null   object 
 4   DateResultRelease  7221 non-null   object 
 5   DateRepConf        7221 non-null   object 
 6   DateDied           7221 non-null   object 
 7   DateRecover        7221 non-null   object 
 8   RemovalType        7221 non-null   object 
 9   Admitted           7221 non-null   object 
 10  RegionRes          7221 non-null   object 
 11  ProvRes            7221 non-null   object 
 12  CityMunRes         7221 non-null   object 
 13  BarangayRes        7221 non-null   object 
 14  HealthStatus       7221 non-null   object 
 15  Quarantined        7221 non-null   object 
 16  DateOnset          72

## Batch 4 data (batch4_manila_copy2) = Finished cleaning


### Total = 7,221 non-null rows

<br>
<br>
<br>

In [50]:
#make a 3rd copy of the dataset
batch4_manila_copy3 = batch4_manila_copy2.copy()

# Drop  columns 'RegionRes','ProvRes', 'CityMunRes'

Not useful since every column only has one value: NCR and CITY OF MANILA

In [51]:
cols = ['RegionRes','ProvRes', 'CityMunRes']
batch4_manila_copy3.drop(columns=cols, inplace=True)

In [52]:
batch4_manila_copy3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7221 entries, 17 to 158543
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Age                7221 non-null   float64
 1   AgeGroup           7221 non-null   object 
 2   Sex                7221 non-null   object 
 3   DateSpecimen       7221 non-null   object 
 4   DateResultRelease  7221 non-null   object 
 5   DateRepConf        7221 non-null   object 
 6   DateDied           7221 non-null   object 
 7   DateRecover        7221 non-null   object 
 8   RemovalType        7221 non-null   object 
 9   Admitted           7221 non-null   object 
 10  BarangayRes        7221 non-null   object 
 11  HealthStatus       7221 non-null   object 
 12  Quarantined        7221 non-null   object 
 13  DateOnset          7221 non-null   object 
 14  Pregnanttab        7221 non-null   object 
 15  ValidationStatus   7221 non-null   object 
dtypes: float64(1), object

In [53]:
batch4_manila_copy3.head()

Age  AgeGroup     Sex  DateSpecimen DateResultRelease DateRepConf  \
17   24.0  20 to 24    MALE  Not Recorded      Not Recorded  2022-11-01   
97   26.0  25 to 29    MALE  Not Recorded      Not Recorded  2022-11-01   
104  54.0  50 to 54    MALE  Not Recorded      Not Recorded  2022-11-01   
121  36.0  35 to 39  FEMALE    2022-10-28        2022-10-29  2022-11-01   
148  22.0  20 to 24    MALE    2022-10-29        2022-10-29  2022-11-01   

         DateDied   DateRecover RemovalType Admitted   BarangayRes  \
17   Not Recorded  Not Recorded   RECOVERED       NO  BARANGAY 675   
97   Not Recorded  Not Recorded   RECOVERED       NO  BARANGAY 459   
104  Not Recorded  Not Recorded   RECOVERED       NO  BARANGAY 318   
121  Not Recorded  Not Recorded   RECOVERED       NO  BARANGAY 179   
148  Not Recorded  Not Recorded   RECOVERED       NO  BARANGAY 654   

    HealthStatus Quarantined     DateOnset Pregnanttab  \
17     RECOVERED          NO  Not Recorded          No   
97     RECOVERED          NO  Not Recorded          No   
104    RECOVERED          NO  Not Recorded          No   
121    RECOVERED          NO  Not Recorded          NO   
148    RECOVERED          NO  Not Recorded          No   

                                      ValidationStatus  
17   Health Status is "Recovered", but no Date Reco...  
97   Health Status is "Recovered", but no Date Reco...  
104  Health Status is "Recovered", but no Date Reco...  
121  Removal Type is "Recovered", but no Recovered ...  
148  Health Status is "Recovered", but no Date Reco...

In [58]:
# Export the cleaned dataset to a CSV file.
batch4_manila_copy3.to_csv('batch4_manila_cleaned2.csv')

In [59]:
batch4_cleaned = pd.read_csv('batch4_manila_cleaned2.csv')


batch4_cleaned.head()

Unnamed: 0   Age  AgeGroup     Sex  DateSpecimen DateResultRelease  \
0          17  24.0  20 to 24    MALE  Not Recorded      Not Recorded   
1          97  26.0  25 to 29    MALE  Not Recorded      Not Recorded   
2         104  54.0  50 to 54    MALE  Not Recorded      Not Recorded   
3         121  36.0  35 to 39  FEMALE    2022-10-28        2022-10-29   
4         148  22.0  20 to 24    MALE    2022-10-29        2022-10-29   

  DateRepConf      DateDied   DateRecover RemovalType Admitted   BarangayRes  \
0  2022-11-01  Not Recorded  Not Recorded   RECOVERED       NO  BARANGAY 675   
1  2022-11-01  Not Recorded  Not Recorded   RECOVERED       NO  BARANGAY 459   
2  2022-11-01  Not Recorded  Not Recorded   RECOVERED       NO  BARANGAY 318   
3  2022-11-01  Not Recorded  Not Recorded   RECOVERED       NO  BARANGAY 179   
4  2022-11-01  Not Recorded  Not Recorded   RECOVERED       NO  BARANGAY 654   

  HealthStatus Quarantined     DateOnset Pregnanttab  \
0    RECOVERED          NO  Not Recorded          No   
1    RECOVERED          NO  Not Recorded          No   
2    RECOVERED          NO  Not Recorded          No   
3    RECOVERED          NO  Not Recorded          NO   
4    RECOVERED          NO  Not Recorded          No   

                                    ValidationStatus  
0  Health Status is "Recovered", but no Date Reco...  
1  Health Status is "Recovered", but no Date Reco...  
2  Health Status is "Recovered", but no Date Reco...  
3  Removal Type is "Recovered", but no Recovered ...  
4  Health Status is "Recovered", but no Date Reco...

In [60]:
batch4_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7221 entries, 0 to 7220
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         7221 non-null   int64  
 1   Age                7221 non-null   float64
 2   AgeGroup           7221 non-null   object 
 3   Sex                7221 non-null   object 
 4   DateSpecimen       7221 non-null   object 
 5   DateResultRelease  7221 non-null   object 
 6   DateRepConf        7221 non-null   object 
 7   DateDied           7221 non-null   object 
 8   DateRecover        7221 non-null   object 
 9   RemovalType        7221 non-null   object 
 10  Admitted           7221 non-null   object 
 11  BarangayRes        7221 non-null   object 
 12  HealthStatus       7221 non-null   object 
 13  Quarantined        7221 non-null   object 
 14  DateOnset          7221 non-null   object 
 15  Pregnanttab        7221 non-null   object 
 16  ValidationStatus   7221 